### PR TITLE
[SYCL] Allow specifying -foffload-lto with the new offload driver and build libdevice with thinLTO

### PR DESCRIPTION
**Pull request by sarnex**
This is the first change in my work on thinLTO for SYCL.

### CODE REVIEW

**Review by bader — COMMENTED** [review]
👍

File: clang/test/Driver/sycl-lto.cpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Verify the usage of -foffload-lto with SYCL.
+// RUN: touch %t.cpp
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/test/Driver/sycl-lto.cpp]: whoops, fixed in latest commit, thx

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -5843,10 +5843,15 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back("-emit-llvm-uselists");
 
     if (IsUsingLTO) {
+      bool IsUsingNewOffloadDriver =
```
[the remark this answers is not in the document]
Comment (sarnex) [reply clang/lib/Driver/ToolChains/Clang.cpp]: fixed in latest commit thx

**Comment by sarnex** [comment]
CUDA fail unrelated https://github.com/intel/llvm/issues/14053

**Review by mdtoguchi — APPROVED** [review]

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 3]
```diff
@@ -91,7 +91,7 @@ function(add_devicelib_obj obj_filename)
 
   set(devicelib-obj-file-new-offload ${obj_new_offload_binary_dir}/${obj_filename}.${new-offload-lib-suffix})
   add_custom_command(OUTPUT ${devicelib-obj-file-new-offload}
-                     COMMAND ${clang} -fsycl -c --offload-new-driver
+                     COMMAND ${clang} -fsycl -c --offload-new-driver -foffload-lto=thin
```
Comment (asudarsa) [inline libdevice/cmake/modules/SYCLLibdevice.cmake:94]: Will this impact the use of these library files in a regular flow (i.e full LTO)?

Thanks

Comment (sarnex) [reply libdevice/cmake/modules/SYCLLibdevice.cmake:94]: I haven't tried but it should not, in that case we read in the full BC file to link. The extra information in the BC file provided added by the thinLTO infrastructure allows linking without loading the full BC file.

File: clang/test/Driver/sycl-lto.cpp [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// Verify the usage of -foffload-lto with SYCL.
+// RUN: not %clangxx -fsycl -foffload-lto=thin %s -### 2>&1 | FileCheck -check-prefix=CHECK_ERROR %s
+//
+// RUN: %clangxx -fsycl --offload-new-driver -foffload-lto=thin %s -### 2>&1 | FileCheck -check-prefix=CHECK_SUPPORTED %s
+
+// CHECK_ERROR: unsupported option '-foffload-lto=thin' for target 'spir64-unknown-unknown'
+
+// CHECK_SUPPORTED: clang{{.*}} "-cc1" "-triple" "spir64-unknown-unknown" {{.*}} "-flto=thin" "-flto-unit"
```
Comment (bader) [inline clang/test/Driver/sycl-lto.cpp:8]: I think usually driver tests are organized like:
```c++
// Test 1
// RUN: ...
// CHECK1: ...
// 
// Test 2
// RUN: ...
// CHECK2: ...
```
This test organization makes it easier to maintain.

**Review by bader — APPROVED** [review]

File: clang/test/Driver/sycl-lto.cpp (hunk 4 above)
Comment (sarnex) [reply clang/test/Driver/sycl-lto.cpp:8]: thanks ill reorder it

**Review by asudarsa — COMMENTED** [review]
Overall looks good except for a question on device library code gen. I also had a high level question. Will it be possible to add a high level design document that will guide this effort?

Thanks

File: libdevice/cmake/modules/SYCLLibdevice.cmake (hunk 3 above)
Comment (asudarsa) [reply libdevice/cmake/modules/SYCLLibdevice.cmake:94]: great. Thanks much

**Comment by sarnex** [comment]
@asudarsa Yes, I plan to add a design document but first we need to resolve some design questions which I will present in a future internal meeting.

**Review by againull — APPROVED** [review]

**Pull request merged**